api/admin/controller.py

```diff
@@ -2410,26 +2425,93 @@ def _sitewide_settings_controller(self, configuration_object):
         protocols = self._get_integration_protocols(provider_apis, protocol_name_attr="NAME")
         protocols.append(dict(name=ExternalIntegration.MANUAL,
                               label=_("Manual import"),
                               description=_("Books will be manually added to the circulation manager, not imported automatically through a protocol."),
                               settings=[],
                               ))
 
+        return protocols
+
+    def _get_prior_test_results(self, collection, protocolClass):
+        """This helper function returns previous self test results for a given
+        collection if it has a protocol.
+        """
+        provider_apis = list(self.PROVIDER_APIS)
+        provider_apis.append(OPDSImportMonitor)
+
+        self_test_results = None
+        protocol = protocolClass
+
+        if not collection or not collection.protocol:
+            return None
+
+        if collection.protocol == OPDSImportMonitor.PROTOCOL:
+            protocol = OPDSImportMonitor
+
+        if protocol in provider_apis and issubclass(protocol, HasSelfTests):
+            if (collection.protocol == OPDSImportMonitor.PROTOCOL):
+                self_test_results = protocol.prior_test_results(self._db, protocol, self._db, collection, OPDSImporter)
+            else:
+                self_test_results = protocol.prior_test_results(self._db, protocol, self._db, collection)
+
+        return self_test_results
+
+    def collection_self_tests(self, identifier):
+        protocols = self._get_collection_protocols(self.PROVIDER_APIS)
+
+        if not identifier:
+            return MISSING_COLLECTION_IDENTIFIER
+
+        if flask.request.method == 'GET':
+            collection = dict()
+            protocolClass = None
+            for col in self._db.query(Collection).filter(Collection.id==int(identifier)):
+                collection = dict(
+                    id=col.id,
+                    name=col.name,
+                    protocol=col.protocol,
+                    parent_id=col.parent_id,
+                    settings=dict(external_account_id=col.external_account_id),
+                )
+
+                if col.protocol in [p.get("name") for p in protocols]:
+                    [protocolClass] = [p for p in self.PROVIDER_APIS if p.NAME == col.protocol]
+
+                collection["self_test_results"] = self._get_prior_test_results(col, protocolClass)
+            return dict(collection=collection)
+
+        if flask.request.method == "POST":
+            collection = dict()
+            collectionProtocol = None
+            protocolClass = None
+            for col in self._db.query(Collection).filter(Collection.id==int(identifier)):
+                collection = col
+                collectionProtocol = col.protocol
+
+                if collectionProtocol in [p.get("name") for p in protocols]:
+                    [protocolClassFound] = [p for p in self.PROVIDER_APIS if p.NAME == col.protocol]
```

Comment:
This doesn't work if you have a manual collection. protocols includes manual but PROVIDER_APIS does not.